### PR TITLE
Serialize JSON directly from Pages

### PR DESCRIPTION
**Pull request by wendigo**
This avoids:
* going Page -> Iterable<List<Object>> -> JSON (instead we are now serializing Page values to JSON directly)
* boxing/unboxing of primitive values (due to a `Object Type.getObjectValue` calls)
* materializing byte[], String from *CharType and VarbinaryType

## Release notes

(x) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
## Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/server/protocol/JsonBytesQueryDataProducer.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+                .orElseThrow(() -> new IllegalStateException("Data present without columns"));
+
+        if (typeEncoders == null) {
+            typeEncoders = createTypeEncoders(session, columns);
+            sourcePageChannels = requireNonNull(columns, "columns is null").stream()
+                    .mapToInt(OutputColumn::sourcePageChannel)
+                    .toArray();
+        }
+
+        // Write to a buffer so we can capture and propage the exception
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/server/protocol/JsonBytesQueryDataProducer.java:58]: propage typo

File: core/trino-main/src/main/java/io/trino/server/protocol/JsonEncodingUtils.java [hunk 2]
```diff
@@ -427,5 +443,20 @@ public void encode(JsonGenerator generator, ConnectorSession session, Block bloc
                 default -> generator.writePOJO(value);
             }
         }
+
+        private Object roundParametricTypes(Object value)
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/server/protocol/JsonEncodingUtils.java:447]: can this PR be more iterative. It is hard to follow what is happening in this commit.

Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/server/protocol/JsonEncodingUtils.java:447]: I've tried to do just that and I failed. I can walk you through it offline

**Review by losipiuk — APPROVED** [review]
looks fine for what i can tell

File: client/trino-client/src/main/java/io/trino/client/ResultRowsDecoder.java [hunk 3]
```diff
@@ -83,8 +83,8 @@ public ResultRows toRows(List<Column> columns, QueryData data)
         }
 
         verify(columns != null && !columns.isEmpty(), "Columns must be set when decoding data");
-        if (data instanceof RawQueryData) {
-            RawQueryData rawData = (RawQueryData) data;
+        if (data instanceof TypedQueryData) {
```
Comment (sajjoseph) [inline client/trino-client/src/main/java/io/trino/client/ResultRowsDecoder.java:86]: can't we short circuit it?
```
if (data instanceof TypedQueryData rawData) {
```

Comment (wendigo) [reply client/trino-client/src/main/java/io/trino/client/ResultRowsDecoder.java:86]: No. This is client so JDK 8
